pyam/plotting.py

```diff
@@ -375,19 +375,103 @@ def stack_plot(df, x='year', y='value', stack='variable',
+    # then cast back to datetime at the end..?
+    _df.index = _df.index.astype(float)
+
+    time_original = _df.index.values
+    first_zero_times = pd.DataFrame(index=["first_zero_time"])
+    both_positive_and_negative = _df.apply(
+        lambda x: (x >= 0).any() and (x < 0).any()
+    )
+
+    for col in _df.loc[:, both_positive_and_negative]:
+        values = _df[col].values
+        for i, val in enumerate(values[:-1]):
+            if np.sign(val) == -1 * np.sign(values[i + 1]):
+                x_1 = time_original[i]
+                x_2 = time_original[i + 1]
+                y_1 = val
+                y_2 = values[i + 1]
+
+                first_zero_time = x_1 - y_1 * (x_2 - x_1) / (y_2 - y_1)
+                first_zero_times.loc[:, col] = first_zero_time
+                if first_zero_time not in _df.index.values:
+                    _df.loc[first_zero_time, :] = np.nan
+
+    first_zero_times = first_zero_times.sort_values(
+        by="first_zero_time",
+        axis=1,
+    )
+    first_zero_times
+    _df = _df.reindex(sorted(_df.index)).interpolate(method="values")
+
+    # Sort lines so that negative timeseries are on the right,
+    # positive timeseries are on the left and timeseries which
+    # go from positive to negative are ordered such that the
+    # timeseries which goes negative first is on the right
+    # (case of timeseries which go from negative to positive
+    # is an edge case we haven't thought about as it's unlikely
+    # to apply to us).
+    col_order = [
+        c for c in first_zero_times.columns[::-1]
+    ]
+    for col in _df:
+        if (_df[col] > 0).all():
+            col_order.insert(0, col)
+        elif (_df[col] < 0).all():
+            col_order.append(col)
+        else:  # already covered by `zero_times` sorting
+            continue
+
+    _df = _df[col_order]
+
     # explicitly get colors
     defaults = default_props(reset=True, num_colors=len(_df.columns),
                              colormap=cmap)['color']
     rc = run_control()
-    colors = []
+    colors = {}
     for key in _df.columns:
         c = next(defaults)
-        if 'color' in rc and stack in rc['color'] and\
-                key in rc['color'][stack]:
+        c_in_rc = 'color' in rc
+        if c_in_rc and stack in rc['color'] and key in rc['color'][stack]:
             c = rc['color'][stack][key]
-        colors.append(c)
-
-    ax.stackplot(_df.index, _df.T, labels=_df.columns, colors=colors, **kwargs)
+        colors[key] = c
+
+    negative_only_cumulative = _df.applymap(
+        lambda x: x if x < 0 else 0
+    ).cumsum(axis=1)
+    positive_only_cumulative = _df.applymap(lambda x: x if x >= 0 else 0)[
+        col_order[::-1]
+    ].cumsum(axis=1)[
+        col_order
+    ]
+
+    total_kwargs = kwargs.pop("total_kwargs", {})
+    plt_total = kwargs.pop("total", False)
+    time = _df.index.values
+    upper = positive_only_cumulative.iloc[:, 0].values
+    for j, col in enumerate(_df):
```

Comment:
I'm having a little trouble understanding what is going on here, could you explain and I can continue the review?

Reply:
Ah ok yep this bit is tricky, mainly because making sure the top and bottom are in the right spot is hard (by default matplotlib doesn't put them in the right spot for this application).

In line 454 we set our upper limit for the first segment to plot. That is simply the sum of the positive only cumulative values  i.e. your upper limit should be either zero, if there are no positive contributions, or more, if there are positive values (cumulative is probably misleading here as it's really the sum of timeseries rather than a cumulative sum over time for a given variable...).

As we loop over the columns of `_df`, which has been sorted from positive to negative, we are basically plotting from the top line (the sum of all positive contributions) down. Hence we next need to find where the lower line for this particular section of the stack should be. 

The starting point for that is `poc_nr` i.e. the sum of all positive contributions up until the next row (maybe we should swap this order in the code). However, if this row includes some negative values, then we need to make sure they show up too. That's what line 462 does: it ensures that if this row has negative values, the lower line will go to those negative values rather than getting stuck at the lower limit of the sum of positive values only i.e. zero. 

Then we have both our upper and lower limits so we can use `fill_between` before finally setting `upper = lower` so that the next segment to plot will appear in the right place. 

Hopefully that helps a bit.

Reply:
Thanks! I think I get it now